### PR TITLE
tests: Add bluetooth/host/keys bt_foreach_bond()

### PR DESCRIPTION
**Pull request by ahmedmoheb-nordic**
This UT project for bt_foreach_bond() under bluetooth/host/keys.c

Regarding the first 2 commits (#ee8cda8cba, #f18d1c4e97), this branch is based on an approve PR [Add bluetooth host keys unit test](https://github.com/zephyrproject-rtos/zephyr/pull/48676)

### CODE REVIEW

**Comment by jori-nordic** [comment]
parent PR is merged, please rebase

**Comment by ahmedmoheb-nordic** [comment]
> parent PR is merged, please rebase

Done!

**Comment by ahmedmoheb-nordic** [comment]
> parent PR is merged, please rebase

Done!

**Review by jhedberg — DISMISSED** [review]

File: tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/* Callback to be used when no calls are expected by bt_foreach_bond() */
+static void bt_foreach_bond_unreachable_cb(const struct bt_bond_info *info, void *user_data)
+{
+	zassert_unreachable("Unexpected call to '%s()' occurred", __func__);
+}
+
+/* Callback to be used when calls are expected by bt_foreach_bond() */
+static void bt_foreach_bond_expected_cb(const struct bt_bond_info *info, void *user_data)
```
Comment (jori-nordic) [inline tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c:35]: could you place the callbacks right before the test case that uses them? it's easier to read without jumping around in the editor.

File: tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	clear_key_pool();
+	int rv = fill_key_pool_by_id_addr(testing_id_addr_pair_lut,
+					  ARRAY_SIZE(testing_id_addr_pair_lut), returned_keys_refs);
+
+	zassert_true(rv == 0, "Failed to fill keys pool list, error code %d", -rv);
+
+	return NULL;
+}
+
+ZTEST_SUITE(bt_keys_foreach_bond_keys_type_not_set, NULL, type_not_set_ts_setup, NULL, NULL, NULL);
```
Comment (jori-nordic) [inline tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c:57]: the two test suites can be merged IMO (e.g. name `bt_keys_foreach_bond`). You can move the setup from test suite setup to test case setup.

File: tests/bluetooth/host/keys/bt_keys_foreach_bond/src/test_suite_foreach_bond_invalid_inputs.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ *  Test if the callback will be called if the ID doesn't exist
+ *
+ *  Constraints:
+ *   - ID doesn't exist in the list
+ *   - Valid callback is passed to bt_keys_foreach_bond()
+ *
+ *  Expected behaviour:
+ *   - Callback should never be called
+ */
+ZTEST(bt_keys_foreach_bond_invalid_inputs, test_callback_non_existing_id)
```
Comment (jori-nordic) [inline tests/bluetooth/host/keys/bt_keys_foreach_bond/src/test_suite_foreach_bond_invalid_inputs.c:50]: can you add another one checking no callbacks are called when the keys pool/list is empty (but the other params are valid)?

**Review by jori-nordic — CHANGES_REQUESTED** [review]
a few comments, but otherwise ok

File: tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c (hunk 2 above)
Comment (jori-nordic) [reply tests/bluetooth/host/keys/bt_keys_foreach_bond/src/main.c:39]: no, the test case function should be executed once before each test case, that's the whole point. It's the test suite function that is executed once per suite.

**Review by jori-nordic — DISMISSED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**